drivers/cache/cache_aspeed.c

```diff
@@ -0,0 +1,267 @@
+ *	bit[22]: 23th 32KB from 0x000a_8000 to 0x000a_ffff
+ *	bit[23]: 24th 32KB from 0x000b_0000 to 0x000b_ffff
+ */
+#define CACHE_AREA_CTRL_REG	0xa50
+#define CACHE_INVALID_REG	0xa54
+#define CACHE_FUNC_CTRL_REG	0xa58
+
+#define CACHED_SRAM_ADDR	CONFIG_SRAM_BASE_ADDRESS
+#define CACHED_SRAM_SIZE	CONFIG_SRAM_SIZE
+#define CACHED_SRAM_END		(CACHED_SRAM_ADDR + CACHED_SRAM_SIZE - 1)
+
+#define CACHE_AREA_SIZE_LOG2	15
+#define CACHE_AREA_SIZE		(1 << CACHE_AREA_SIZE_LOG2)
+
+#define DCACHE_INVALID(addr)	(BIT(31) | ((addr & GENMASK(10, 0)) << 16))
+#define ICACHE_INVALID(addr)	(BIT(15) | ((addr & GENMASK(10, 0)) << 0))
+
+#define ICACHE_CLEAN		BIT(2)
+#define DCACHE_CLEAN		BIT(1)
+#define CACHE_EANABLE		BIT(0)
+
+/* cache size = 32B * 128 = 4KB */
+#define CACHE_LINE_SIZE_LOG2	5
+#define CACHE_LINE_SIZE		(1 << CACHE_LINE_SIZE_LOG2)
+#define N_CACHE_LINE		128
+#define CACHE_ALIGNED_ADDR(addr) \
+	((addr >> CACHE_LINE_SIZE_LOG2) << CACHE_LINE_SIZE_LOG2)
+
+/* prefetch buffer */
+#define PREFETCH_BUF_SIZE	CACHE_LINE_SIZE
+
+static void aspeed_cache_init(void)
+{
+	uint32_t base = DT_REG_ADDR(DT_NODELABEL(syscon));
+	uint32_t start_bit, end_bit;
+
+	/* set all cache areas to no-cache by default */
+	sys_write32(0, base + CACHE_FUNC_CTRL_REG);
+
+	/* calculate how many areas need to be set */
+	start_bit = CACHED_SRAM_ADDR >> CACHE_AREA_SIZE_LOG2;
+	end_bit = CACHED_SRAM_END >> CACHE_AREA_SIZE_LOG2;
+	sys_write32(GENMASK(end_bit, start_bit), base + CACHE_AREA_CTRL_REG);
+
+	/* enable cache */
+	sys_write32(CACHE_EANABLE, base + CACHE_FUNC_CTRL_REG);
+}
+
+/**
+ * @brief get aligned address and the number of cachline to be invalied
+ * @param [IN] addr - start address to be invalidated
+ * @param [IN] size - size in byte
+ * @param [OUT] p_aligned_addr - pointer to the cacheline aligned address variable
+ * @return number of cacheline to be invalidated
+ *
+ *  * addr
+ *   |--------size-------------|
+ * |-----|-----|-----|-----|-----|
+ *  \                             \
+ *   head                          tail
+ *
+ * example 1:
+ * addr = 0x100 (cacheline aligned), size = 64
+ * then head = 0x100, number of cache line to be invalidated = 64 / 32 = 2
+ * which means range [0x100, 0x140) will be invalidated
+ *
+ * example 2:
+ * addr = 0x104 (cacheline unaligned), size = 64
+ * then head = 0x100, number of cache line to be invalidated = 1 + 64 / 32 = 3
+ * which means range [0x100, 0x160) will be invalidated
+ */
+static uint32_t get_n_cacheline(uint32_t addr, uint32_t size, uint32_t *p_head)
+{
+	uint32_t n = 0;
+	uint32_t tail;
+
+	/* head */
+	*p_head = CACHE_ALIGNED_ADDR(addr);
+
+	/* roundup the tail address */
+	tail = addr + size + (CACHE_LINE_SIZE - 1);
+	tail = CACHE_ALIGNED_ADDR(tail);
+
+	n = (tail - *p_head) >> CACHE_LINE_SIZE_LOG2;
+
+	return n;
+}
+
+void cache_data_enable(void)
+{
+	aspeed_cache_init();
+}
+
+void cache_data_disable(void)
+{
+	uint32_t base = DT_REG_ADDR(DT_NODELABEL(syscon));
+
+	sys_write32(0, base + CACHE_FUNC_CTRL_REG);
```

Comment:
you cannot really use `sys_write32` & co to interface with syscon. We have in place a syscon API for that. Look at `include/zephyr/drivers/syscon.h`. You should use `syscon_get_base()`, `syscon_read_reg()` & co.

Reply:
Thank you for your comment.  I've added a commit to enable generic syscon driver and used syscon api to access syscon registers.